### PR TITLE
Pairwise postprocessing in validation for both: images and representations

### PR DESCRIPTION
**Pull request by AlekseySh**

### CODE REVIEW

File: oml/retrieval/postprocessors/pairwise.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        n_galleries = len(galleries)
+
+        assert list(distances.shape) == [n_queries, n_galleries]
+
+        # 1. Adjust top_n with respect to the actual gallery size and find top-n pairs
+        top_n = min(self.top_n, n_galleries)
+        ii_top = torch.topk(distances, k=top_n, largest=False)[1].view(-1)
+
+        # 2. Create (n_queries * top_n) pairs of each query and related galleries and re-estimate distances for them
+        distances_upd = self.inference(queries=queries, galleries=galleries, ii_top=ii_top, top_n=top_n)
```
Comment (dapladoc) [inline oml/retrieval/postprocessors/pairwise.py:53]: Here `ii_top` is a flattened array, but in the `self.inference` docstring it is said that `ii_top: Indices of the closest galleries with the shape of [Q, top_n]`. Is it ok to use flattened array in such case?

File: oml/retrieval/postprocessors/pairwise.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            galleries: Galleries representations with the length of ``G``
+            ii_top: Indices of the closest galleries with the shape of ``[Q, top_n]``
+            top_n: Defines the number of the closest galleries to re-rank
+
+        Returns:
+            Updated distance matrix with the shape of ``[Q, G]``.
+
+        """
+        queries = queries.repeat_interleave(top_n, dim=0)
+        galleries = galleries[ii_top]
```
Comment (dapladoc) [inline oml/retrieval/postprocessors/pairwise.py:137]: If `ii_top` is of size `[Q, top_n]` then `galleries` will be of size `[Q, top_n, dim]`, where `dim` is the embeddings dimension. It doesn't look like it is supposed to be.

File: oml/inference/base.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+) -> Tensor:
+    if f_imread is None:
+        f_imread = get_im_reader_for_transforms(transform)
+
+    prev_mode = model.training
+    model.eval()
+
+    dataset = ListDataset(paths, bboxes=None, transform=transform, f_imread=f_imread)
+    loader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+    device = next(model.parameters()).device
```
Comment (dapladoc) [inline oml/inference/base.py:32]: The same thoughts here about introducing the `get_device` function, and the context manager to restore model training flag.

**Comment by AlekseySh** [comment]
* added pairwise models (and corresponding pair datasets + inference code for them)
* added pairwise postprocessors (on embeddings and on images). organised code, so it's shared among both postprocessors 
* Added new models and postprocessors to the registry and docs
* added creating a postprocessor to the validation entry point (not to a training one yet)
* Added tests for postprocessor on images pairs

File: oml/inference/pairwise.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from oml.utils.images.images import TImReader
+
+
+def pairwise_inference(
+    model: IPairwiseDistanceModel, dataset: IPairsDataset, num_workers: int, batch_size: int, verbose: bool
+) -> Tensor:
+    prev_mode = model.training
+    model.eval()
+    loader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+    device = next(model.parameters()).device
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/inference/pairwise.py:28]: done

File: oml/inference/pairwise.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+# fmt: off
+def pairwise_inference(
+        model: IPairwiseDistanceModel,
+        dataset: IPairsDataset,
+        num_workers: int,
+        batch_size: int,
+        verbose: bool
+) -> Tensor:
+    prev_mode = model.training
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/inference/pairwise.py:25]: The idea is good, but this PR is too big already (~800 lines). So, if you or someone else do it as a small separate PR I will be happy

File: oml/retrieval/postprocessors/pairwise.py (hunk 1 above)
Comment (AlekseySh) [reply oml/retrieval/postprocessors/pairwise.py:53]: thx for your findings 
code is okay, but I will fix comments

File: oml/retrieval/postprocessors/pairwise.py (hunk 2 above)
Comment (AlekseySh) [reply oml/retrieval/postprocessors/pairwise.py:137]: thx for your findings 
code is okay, but I will fix comments

File: oml/retrieval/postprocessors/pairwise.py (hunk 1 above)
Comment (AlekseySh) [reply oml/retrieval/postprocessors/pairwise.py:53]: thx for your findings
I've aligned shapes and comments for `ii_top` and `distances`

File: oml/retrieval/postprocessors/pairwise.py (hunk 2 above)
Comment (AlekseySh) [reply oml/retrieval/postprocessors/pairwise.py:137]: thx for your findings
I've aligned shapes and comments for `ii_top` and `distances`

File: oml/inference/base.py (hunk 3 above)
Comment (AlekseySh) [reply oml/inference/base.py:32]: done

File: oml/metrics/embeddings.py [hunk 6]
```diff
@@ -170,17 +172,16 @@ def _calc_matrices(self) -> None:
         if self.postprocessor is None:
             pass
         elif isinstance(self.postprocessor, PairwiseEmbeddingsPostprocessor):
-            max_k = max([*self.cmc_top_k, *self.precision_top_k, *self.map_top_k])
-            if max_k > self.postprocessor.top_n:
-                warn(
-                    f"One of retrieval metrics will be computed at k = {max_k},"
-                    f"but postprocessor will re-rank only {self.postprocessor.top_n} closest galleries."
-                    f"Make sure that this is the desired behaviour."
-                )
```
Comment (dapladoc) [inline oml/metrics/embeddings.py:179]: Sorry, I can't find where you decided to put this warning. Or did you get rid of them at all?

File: oml/inference/pairwise.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/inference/pairwise.py:26]: okay, I added it in this PR

[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/inference/pairwise.py:26]: thx for a draft

File: docs/source/contents/postprocessing.rst [hunk 7]
```diff
@@ -1,48 +1,108 @@
...
+IPairsDataset
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. autoclass:: oml.interfaces.datasets.IPairsDataset
+    :undoc-members:
+    :show-inheritance:
+
+    .. automethod:: __init__
+    .. automethod:: __getitem__
+
+Implementations
```
Comment (DaloroAT) [inline docs/source/contents/postprocessing.rst:39]: readthedocs looks ugly, it should be nested

File: oml/inference/base.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+# todo: use lightning for half precision and DDP
+@torch.no_grad()
+def inference_on_images(
+    model: nn.Module,
+    paths: List[Path],
+    transform: TTransforms,
+    num_workers: int,
+    batch_size: int,
+    verbose: bool,
```
Comment (DaloroAT) [inline oml/inference/base.py:23]: Suppose `verbose: bool = False` is ok

File: oml/inference/base.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+from oml.utils.misc_torch import get_device, temporary_setting_model_mode
+
+
+# todo: use lightning for half precision and DDP
+@torch.no_grad()
+def inference_on_images(
+    model: nn.Module,
+    paths: List[Path],
+    transform: TTransforms,
+    num_workers: int,
```
Comment (DaloroAT) [inline oml/inference/base.py:21]: `num_workers: int = 0` is also ok

File: oml/metrics/embeddings.py [hunk 10]
```diff
@@ -170,17 +172,16 @@ def _calc_matrices(self) -> None:
...
                 distances=self.distance_matrix,
-                emb_query=embeddings[is_query],  # type: ignore
-                emb_gallery=embeddings[is_gallery],  # type: ignore
+                queries=embeddings[is_query],  # type: ignore
+                galleries=embeddings[is_gallery],  # type: ignore
+            )
+        elif isinstance(self.postprocessor, PairwiseImagesPostprocessor):
+            self.distance_matrix = self.postprocessor.process(
+                distances=self.distance_matrix,
+                queries=np.array(self.acc.storage[PATHS_KEY])[is_query],  # type: ignore
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/metrics/embeddings.py:183]: but we don't know in advance if our postprocessor need this key or not
I suggest to do it in the next pr when we introduce method `postprocessor.get_needed_keys()`

File: oml/registry/pairwise_models.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+from oml.const import TCfg
+from oml.interfaces.models import IPairwiseModel
+from oml.models.siamese import LinearSiamese, ResNetSiamese
+from oml.utils.misc import dictconfig_to_dict
+
+PAIRWISE_MODELS_REGISTRY = {"linear_siamese": LinearSiamese, "resnet_siamese": ResNetSiamese}
+
+
+def get_pairwise_model(model_name: str, **kwargs: Dict[str, Any]) -> IPairwiseModel:
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/registry/pairwise_models.py:11]: fixed

File: oml/utils/misc_torch.py [hunk 12]
```diff
@@ -95,6 +96,16 @@ def pairwise_dist(x1: Tensor, x2: Tensor, p: int = 2) -> Tensor:
     return cdist(x1=x1, x2=x2, p=p)
 
 
+def normalise(x: Tensor, p: int = 2) -> Tensor:
+    xn = torch.linalg.norm(x, p, dim=1).detach()
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/utils/misc_torch.py:100]: added assert and docstring

File: oml/utils/misc_torch.py [hunk 13]
```diff
@@ -103,6 +114,14 @@ def _check_is_sequence(val: Any) -> bool:
...
+def temporary_setting_model_mode(model: torch.nn.Module, set_train: bool) -> torch.nn.Module:
+    prev_mode = model.training
+    model.train(set_train)
+    yield model
+    model.train(prev_mode)
+
+
 class OnlineCalc(ABC):
     """
     The base class to calculate some statistics online (on the steam of values).
```
[the remark this answers is not in the document]
Comment (AlekseySh) [reply oml/utils/misc_torch.py:127]: fixed

File: oml/metrics/embeddings.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (DaloroAT) [reply oml/metrics/embeddings.py:183]: I mean that you need to introduce `needed_keys`, because right now it's hardcoded and you will fail only all computations

Ok, next PR

File: oml/inference/base.py (hunk 8 above)
Comment (AlekseySh) [reply oml/inference/base.py:23]: agree

File: oml/inference/base.py (hunk 9 above)
Comment (AlekseySh) [reply oml/inference/base.py:21]: I prefer to specify it directly because it's easy to forget
And then you may need a lot of time to notice that inference takes too long

File: docs/source/contents/postprocessing.rst (hunk 7 above)
Comment (AlekseySh) [reply docs/source/contents/postprocessing.rst]: reworked

**Review by DaloroAT — APPROVED** [review]
